### PR TITLE
Using the new Comm Python package

### PR DESCRIPTION
**Pull request by martinRenou**
Related discussions:
- https://github.com/jupyter-xeus/xeus-python/issues/342 
- https://github.com/jupyter-widgets/ipywidgets/issues/3514

Currently working on a Python comm package on https://github.com/martinRenou/comm for now, but it should probably be moved in the IPython org

This has the benefit of being able to:
- remove the ipykernel dependency in ipywidgets https://github.com/jupyter-widgets/ipywidgets/pull/3533
- remove the ipykernel monkey patch in xeus-python https://github.com/jupyter-xeus/xeus-python/pull/548
- remove the ipykernel package mock in jupyterlite's pyolite 
- remove the ipykernel package mock in emscripten-forge

### CODE REVIEW

File: python/ipywidgets/ipywidgets/__init__.py [hunk 1]
```diff
@@ -21,28 +21,26 @@
...
+
 
 def load_ipython_extension(ip):
     """Set up Jupyter to work with widgets"""
-    if not hasattr(ip, 'kernel'):
-        return
-    register_comm_target(ip.kernel)
+    register_comm_target()
 
-def register_comm_target(kernel=None):
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/__init__.py:34]: should we keep the argument for backwards compatibility?

**Comment by SylvainCorlay** [comment]
> This has the benefit of being able to:
> 
> remove the ipykernel dependency in ipywidgets https://github.com/jupyter-widgets/ipywidgets/pull/3533
> remove the ipykernel monkey patch in xeus-python https://github.com/jupyter-xeus/xeus-python/pull/548
> remove the ipykernel package mock in jupyterlite's pyolite
> remove the ipykernel package mock in emscripten-forge

Wow. I think it is quite an improvement.

File: python/ipywidgets/ipywidgets/__init__.py [hunk 2]
```diff
@@ -21,28 +21,36 @@
...
 
 def load_ipython_extension(ip):
     """Set up Jupyter to work with widgets"""
     if not hasattr(ip, 'kernel'):
         return
-    register_comm_target(ip.kernel)
+    register_comm_target()
 
-def register_comm_target(kernel=None):
+def register_comm_target():
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/__init__.py:42]:
```suggestion
def register_comm_target(kernel=None):
```
I think we shouldn't change the API, as that would be a breaking change.

File: python/ipywidgets/ipywidgets/widgets/tests/utils.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
 # Copyright (c) Jupyter Development Team.
 # Distributed under the terms of the Modified BSD License.
 
-from ipykernel.comm import Comm
 from ipywidgets import Widget
 import ipywidgets.widgets.widget
 
-class DummyComm(Comm):
+import comm
+from ipykernel.comm import Comm
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/widgets/tests/utils.py:8]:
```suggestion
```
import is not used right?

File: python/ipywidgets/ipywidgets/widgets/tests/utils.py [hunk 4]
```diff
@@ -16,16 +19,30 @@ def __init__(self, *args, **kwargs):
...
+def dummy_get_comm_manager(**kwargs):
+    return {}
+
+
 _widget_attrs = {}
 undefined = object()
 
 def setup_test_comm():
+    comm.create_comm = dummy_create_comm
+    comm.get_comm_manager = dummy_get_comm_manager
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/widgets/tests/utils.py:45]: should we restore this? in a try/finally?

File: python/ipywidgets/ipywidgets/widgets/widget_output.py [hunk 5]
```diff
@@ -134,15 +132,6 @@ def __exit__(self, etype, evalue, tb):
...
                 ip.showtraceback((etype, evalue, tb), tb_offset=0)
-            elif self.comm is not None and self.comm.kernel is not None:
-                kernel = self.comm.kernel
-                kernel.send_response(kernel.iopub_socket,
-                                     u'error',
-                                     {
-                    u'traceback': ["".join(traceback.format_exception(etype, evalue, tb))],
-                    u'evalue': repr(evalue.args),
-                    u'ename': etype.__name__
-                    })
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/widgets/widget_output.py:145]: Why can we remove this?

**Review by maartenbreddels — COMMENTED** [review]
Great work.

I do worry we do not have a test yet that exposes https://github.com/ipython/ipykernel/issues/1090

I'm pretty sure this fixes it, but #3653 also didn't pick up this issue, which leaves me a bit worries that we're vulnerable for regressions.

File: python/ipywidgets/ipywidgets/widgets/tests/utils.py (hunk 4 above)
Comment (martinRenou) [reply python/ipywidgets/ipywidgets/widgets/tests/utils.py:45]: We should 👍🏽 I will restore them in the teardown function

File: python/ipywidgets/ipywidgets/widgets/tests/utils.py (hunk 3 above)
Comment (martinRenou) [reply python/ipywidgets/ipywidgets/widgets/tests/utils.py:8]: It is used later in that file, in the teardown function for restoring the original Comm class

**Comment by martinRenou** [comment]
Thanks a lot for your review @maartenbreddels ! I will resolve your comment and get the CI green hopefully :D

File: python/ipywidgets/ipywidgets/widgets/widget_output.py (hunk 5 above)
Comment (martinRenou) [reply python/ipywidgets/ipywidgets/widgets/widget_output.py:145]: I will bring it back. What is annoying with this code is that it's very specific to ipykernel, but I can add more checks so we don't go through this with *e.g.* xeus-python

**Comment by martinRenou** [comment]
@maartenbreddels it was difficult to get green but now it seems good!

Thankfully, the Python 3.7 build requires an old version of ipykernel (latest ipykernel does not support 3.7?) so we have tests against the old versions and new versions.

**Comment by martinRenou** [comment]
Would you have time to make it a review @maartenbreddels ?

File: python/ipywidgets/ipywidgets/widgets/widget.py [hunk 6]
```diff
@@ -686,7 +693,7 @@ def notify_change(self, change):
         # Send the state to the frontend before the user-registered callbacks
         # are called.
         name = change['name']
-        if self.comm is not None and self.comm.kernel is not None:
+        if self.comm is not None and (self.comm.kernel is not None if hasattr(self.comm, "kernel") else True):
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/widgets/widget.py:696]:
```suggestion
        if self.comm is not None and getattr(self.comm, 'kernel', True) is not None:
```
do you think this look better.. ?

File: python/ipywidgets/ipywidgets/widgets/widget_output.py [hunk 7]
```diff
@@ -111,9 +111,9 @@ def __enter__(self):
         kernel = None
         if ip and getattr(ip, "kernel", None) is not None:
             kernel = ip.kernel
-        elif self.comm is not None and self.comm.kernel is not None:
+        elif self.comm is not None and getattr(self.comm, "kernel", None) is not None:
```
Comment (maartenbreddels) [inline python/ipywidgets/ipywidgets/widgets/widget_output.py:114]:
```suggestion
        elif self.comm is not None and getattr(self.comm, 'kernel', True) is not None:
```
then also this?

**Review by maartenbreddels — COMMENTED** [review]
awesome work, some minor comments

File: python/ipywidgets/ipywidgets/widgets/tests/test_widget_templates.py [hunk 8]
```diff
@@ -229,7 +229,7 @@ def test_update_dynamically(self, send_state): #pylint: disable=no-self-use
         assert box.layout.grid_template_areas == ('"top-left top-right"\n' +
                                                   '"bottom-left bottom-right"')
         # check whether frontend was informed
-        send_state.assert_called_once_with(key="grid_template_areas")
+        send_state.assert_called_with(key="grid_template_areas")
```
[the remark this answers is not in the document]
Comment (martinRenou) [reply python/ipywidgets/ipywidgets/widgets/tests/test_widget_templates.py:232]: Yes, I think the test was faulty from the beginning. Though it's unclear to me as to why my PR fixes it.

Setting `bottom_left` would not only trigger [changing grid_template_areas](https://github.com/jupyter-widgets/ipywidgets/blob/6f6156c7e2421f437614b0649a5a3f37c3b4a3ec/python/ipywidgets/ipywidgets/widgets/widget_templates.py#L448) but also [a couple more properties from layout](https://github.com/jupyter-widgets/ipywidgets/blob/6f6156c7e2421f437614b0649a5a3f37c3b4a3ec/python/ipywidgets/ipywidgets/widgets/widget_templates.py#L446-L447)

**Comment by martinRenou** [comment]
Thanks for the review! I resolved your comments

**Comment by maartenbreddels** [comment]
@martinRenou do you know what caused that change?
I'm fine removing the newlines, but maybe good to be aware what is responsible for this change?



> Hello. Is there a plan to merge and release this soon?

Yes, I think we should get this released asap, which I could do tomorrow if we manage to get this in.

**Comment by martinRenou** [comment]
> Yes, I think we should get this released asap, which I could do tomorrow if we manage to get this in.

💯

**Comment by martinRenou** [comment]
@maartenbreddels all green!

**Comment by maartenbreddels** [comment]
@martinRenou great work, i guess we also need to backport this to 7.x
do you want to try the bot?

**Comment by martinRenou** [comment]
@meeseeksdev please backport to 7.x

**Comment by bollwyvl** [comment]
If seeing errors like this in `pytest`:

```py
    def _create_comm(*args, **kwargs):
        """Create a Comm.
    
        This method is intended to be replaced, so that it returns your Comm instance.
        """
>       raise NotImplementedError("Cannot ")
E       NotImplementedError: Cannot
```

Adding:

```py
import ipykernel.ipkernel  # noqa
```

Is enough to trigger the [side-effects](https://github.com/ipython/ipykernel/blob/a895ce72e4c0ecff95280e68ae96faa283e561cf/ipykernel/ipkernel.py#L66) to overload that manager.

Seems like it should go in the docs... or even in that unhelpful traceback itself, cuz 99% of the time, i reckon people are testing with `ipykernel` lying around.

**Pull request merged**